addons/l10n_in/static/src/components/hsn_autocomplete/hsn_autocomplete.js

```diff
@@ -0,0 +1,114 @@
+/** @odoo-module **/
+
+import { AutoComplete } from "@web/core/autocomplete/autocomplete";
+import { useChildRef } from "@web/core/utils/hooks";
+import { registry } from "@web/core/registry";
+import { _t } from "@web/core/l10n/translation";
+import { CharField, charField } from "@web/views/fields/char/char_field";
+import { useInputField } from "@web/views/fields/input_field_hook";
+
+const l10N_IN_HSN_SERVICE_URL = "https://services.gst.gov.in/commonservices/hsn/search/qsearch";
+
+export class L10nInHsnAutoComplete extends CharField {
+    static template = "l10n_in.hsnAutoComplete";
+    static components = {
+        ...CharField.components,
+        AutoComplete,
+    };
+    static props = {
+        ...CharField.props,
+        l10nInHsnDescription: { type: String, optional: true },
+    };
+
+    setup() {
+        super.setup();
+        this.inputRef = useChildRef();
+        useInputField({
+            getValue: () => this.props.record.data[this.props.name] || "",
+            parse: (v) => this.parse(v),
+            ref: this.inputRef,
+        });
+    }
+
+    async getHsnSuggestions(value) {
+        const suggestions = [];
+        const onlyDigits = !isNaN(value) && value.indexOf(" ") < 0;
+        const params = [
+            { type: "byCode", category: "null" }, // For code
+            { type: "byDesc", category: "P" }, // For products
+            { type: "byDesc", category: "S" }, // For services
+        ];
+        const filteredParams = onlyDigits ? [params[0]] : params.slice(1);
+        try {
+            await Promise.all(
+                filteredParams.map(async (param) => {
+                    const res = await fetch(
```

Comment:
In a fetch request you can use an `AbortController` [1] and set a custom timeout.
Somethings like this

[1] https://developer.mozilla.org/en-US/docs/Web/API/AbortController

```suggestion
                    const controller = new AbortController();
                    const signal = controller.signal;
                    setTimeout(() => controller.abort(), 5000);
                    const res = await fetch(url, { signal });
```